man/dmimport.1
.TH "DMIMPORT" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmimport \(en imports records from file into database
.SH SYNOPSIS
\f[B]dmimport\f[R] \-\-help
.PP
\f[B]dmimport\f[R] \-\-version
.PP
\f[B]dmimport\f[R] \-\-\f[B]input\f[R] \f[I]file\f[R]
\-\-\f[B]database\f[R] \f[I]file\f[R] \-\-\f[B]type\f[R] \f[I]type\f[R]
[\-\-\f[B]quote\f[R] \f[I]char\f[R]] [\-\-\f[B]separator\f[R]
\f[I]char\f[R]] [\-\-\f[B]dry\f[R]] [\-\-\f[B]verbose\f[R]]
.SH DESCRIPTION
The \f[B]dmimport\f[R] program reads logs, nodes, sensors, targets, and
observations in CSV format from file and imports them into the database.
The database inserts are transaction\-based.
If an error occurs, the transaction is rolled back, and no records are
written into the database at all.
.PP
The database has to be a valid DMPACK database and must contain the
tables required for the input records.
The nodes, sensors, and targets referenced by input observations must
exist in the database.
The nodes referenced by input sensors must exist as well.
.SH OPTIONS
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
Path of the SQLite database (required, unless in dry mode).
.TP
\f[B]\-\-dry\f[R], \f[B]\-D\f[R]
Enable dry mode.
Read records from file and validate them without actual database import.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-input\f[R], \f[B]\-i\f[R] \f[I]file\f[R]
Path of input file in CSV format.
.TP
\f[B]\-\-quote\f[R], \f[B]\-q\f[R] \f[I]char\f[R]
CSV quote character.
If not passed, quoting is disabled by default.
.TP
\f[B]\-\-separator\f[R], \f[B]\-s\f[R] \f[I]char\f[R]
CSV separator character.
.TP
\f[B]\-\-type\f[R], \f[B]\-t\f[R] [log|node|observ|sensor|target]
Type of records to import (required).
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print progress to \f[I]stdout\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Records have been imported.
.TP
\f[B]1\f[R]
Failure.
Import of records failed.
.SH EXAMPLE
Import observations from CSV file into database:
.IP
.EX
$ dmimport \-t observ \-i observ.csv \-d observ.db \-V
.EE
.SH SEE ALSO
\f[I]dmexport(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmimport.f90
! dmimport.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmimport
    !! Imports nodes, observations, sensors, targets, and logs from CSV file
    !! into database.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmimport'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_type
        !! Command-line arguments.
        character(FILE_PATH_LEN) :: database  = ' '       !! Path to database.
        character(FILE_PATH_LEN) :: input     = ' '       !! Input file path.
        integer                  :: type      = TYPE_NONE !! Entity type.
        character                :: quote     = ASCII_NUL !! CSV quote character (disabled by default).
        character                :: separator = ','       !! CSV separator character.
        logical                  :: dry       = .false.   !! Dry run.
        logical                  :: verbose   = .false.   !! Print progress to standard output.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App type.

    ! Initialise DMPACK.
    call dm_init()

    ! Read command-line arguments.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Create selected database type.
    rc = import(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    integer function import(app) result(rc)
        integer(i8), parameter :: PROGRESS_STEP_SIZE = 500_i8 !! Import progress step size.

        type(app_type), intent(inout) :: app

        integer          :: error, stat, unit
        integer(i8)      :: nrecs, nrows
        logical          :: has, valid
        real(r8)         :: dt
        type(db_type)    :: db
        type(timer_type) :: timer

        type(log_type)    :: log
        type(node_type)   :: node
        type(observ_type) :: observ
        type(sensor_type) :: sensor
        type(target_type) :: target

        ! Try to open input file.
        rc = E_IO
        open (action='read', file=trim(app%input), iostat=stat, newunit=unit)

        if (stat /= 0) then
            call dm_error_out(rc, 'failed to open file ' // trim(app%input))
            return
        end if

        if (app%verbose) then
            print '("Opened file ", a)', trim(app%input)
        end if

        import_block: block
            if (.not. app%dry) then
                ! Try to open database.
                rc = dm_db_open(db, app%database, validate=.true.)

                if (rc == E_INVALID) then
                    call dm_error_out(rc, 'invalid database ' // trim(app%database))
                    exit import_block
                end if

                if (dm_is_error(rc)) then
                    call dm_error_out(rc, 'failed to open database ' // trim(app%database))
                    exit import_block
                end if

                if (app%verbose) then
                    print '("Opened database ", a)', trim(app%database)
                end if

                ! Check for appropriate database table.
                select case (app%type)
                    case (TYPE_NODE);   has = dm_db_table_has(db, SQL_TABLE_NODES)
                    case (TYPE_SENSOR); has = dm_db_table_has(db, SQL_TABLE_SENSORS)
                    case (TYPE_TARGET); has = dm_db_table_has(db, SQL_TABLE_TARGETS)
                    case (TYPE_OBSERV); has = dm_db_table_has_observs(db)
                    case (TYPE_LOG);    has = dm_db_table_has_logs(db)
                end select

                if (.not. has) then
                    rc = E_INVALID
                    call dm_error_out(rc, 'database table not found')
                    exit import_block
                end if
            end if

            ! Start timer.
            if (app%verbose) call dm_timer_start(timer)

            if (.not. app%dry) then
                ! Start database transaction.
                rc = dm_db_begin(db)

                if (dm_is_error(rc)) then
                    call dm_error_out(rc, 'database transaction error')
                    exit import_block
                end if
            end if

            nrecs = 0 ! Number of records written to database.
            nrows = 0 ! Number of rows in CSV file.

            ! Read records from file and insert them into database.
            read_loop: do
                nrows = nrows + 1

                ! Read record from file.
                select case (app%type)
                    case (TYPE_NODE);   rc = dm_csv_read(node,   unit, app%separator, app%quote)
                    case (TYPE_SENSOR); rc = dm_csv_read(sensor, unit, app%separator, app%quote)
                    case (TYPE_TARGET); rc = dm_csv_read(target, unit, app%separator, app%quote)
                    case (TYPE_OBSERV); rc = dm_csv_read(observ, unit, app%separator, app%quote)
                    case (TYPE_LOG);    rc = dm_csv_read(log,    unit, app%separator, app%quote)
                end select

                ! Ignore comments and empty rows.
                if (rc == E_EOR) cycle read_loop

                ! End of file reached.
                if (rc == E_EOF) then
                    rc = E_NONE
                    exit read_loop
                end if

                ! Any other error.
                if (dm_is_error(rc)) then
                    call dm_error_out(rc, 'failed to read record in row ' // dm_itoa(nrows))
                    exit read_loop
                end if

                ! Validate record but skip database insert on dry run.
                if (app%dry) then
                    select case (app%type)
                        case (TYPE_NODE);   valid = dm_node_is_valid(node)
                        case (TYPE_SENSOR); valid = dm_sensor_is_valid(sensor)
                        case (TYPE_TARGET); valid = dm_target_is_valid(target)
                        case (TYPE_OBSERV); valid = dm_observ_is_valid(observ)
                        case (TYPE_LOG);    valid = dm_log_is_valid(log)
                    end select

                    if (.not. valid) then
                        rc = E_INVALID
                        call dm_error_out(rc, 'invalid record in row ' // dm_itoa(nrows))
                        exit read_loop
                    end if

                    cycle read_loop
                end if

                ! Validate and insert record.
                select case (app%type)
                    case (TYPE_NODE);   rc = dm_db_insert(db, node)
                    case (TYPE_SENSOR); rc = dm_db_insert(db, sensor)
                    case (TYPE_TARGET); rc = dm_db_insert(db, target)
                    case (TYPE_OBSERV); rc = dm_db_insert(db, observ)
                    case (TYPE_LOG);    rc = dm_db_insert(db, log)
                end select

                ! Handle database result.
                select case (rc)
                    case (E_NONE)
                        nrecs = nrecs + 1
                        if (.not. app%verbose) cycle read_loop
                        if (modulo(nrecs, PROGRESS_STEP_SIZE) == 0) then
                            print '("Imported ", i0, " records")', nrecs
                        end if

                    case (E_INVALID)
                        call dm_error_out(rc, 'invalid record in row ' // dm_itoa(nrows))
                        exit read_loop

                    case (E_DB_CONSTRAINT)
                        call dm_error_out(rc, 'record in row ' // dm_itoa(nrows) // ' already exists in database')
                        exit read_loop

                    case default
                        call dm_error_out(rc, 'failed to insert record in row ' // dm_itoa(nrows))
                        exit read_loop
                end select
            end do read_loop

            if (.not. app%dry) then
                ! Rollback transaction on error.
                if (dm_is_error(rc)) then
                    error = dm_db_rollback(db)

                    if (dm_is_error(error)) then
                        call dm_error_out(error, 'failed to roll back database transaction')
                        exit import_block
                    end if

                    if (app%verbose) print '("All database changes rolled back")'
                    exit import_block
                end if

                ! Commit transaction to database.
                rc = dm_db_commit(db)

                ! Rollback transaction on error.
                if (dm_is_error(rc)) then
                    call dm_error_out(rc, 'failed to commit database transaction')
                    error = dm_db_rollback(db)

                    if (dm_is_error(error)) then
                        call dm_error_out(error, 'failed to roll back database transaction')
                        exit import_block
                    end if

                    if (app%verbose) print '("All database changes rolled back")'
                    exit import_block
                end if
            end if

            ! Output statistics.
            if (app%verbose) then
                call dm_timer_stop(timer, dt)
                print '("Read ", i0, " rows from file ", a)', nrows, trim(app%input)
                print '("Imported ", i0, " records in ", f0.1, " seconds")', nrecs, dt
            end if
        end block import_block

        ! Close database.
        if (.not. app%dry) then
            call dm_db_close(db)
            if (app%verbose) print '("Closed database ", a)', trim(app%database)
        end if

        ! Close file.
        close (unit)
        if (app%verbose) print '("Closed file ", a)', trim(app%input)

        if (dm_is_error(rc)) return
        if (app%verbose) print '("Finished")'
    end function import

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app

        character(6)           :: type_name
        type(arg_parser_class) :: parser

        type_name = ' '

        call parser%add('type',      short='t', type=ARG_TYPE_STRING,   max_len=TYPE_NAME_LEN, required=.true.) ! -t, --type <string>
        call parser%add('database',  short='d', type=ARG_TYPE_DATABASE, required=.true., exist=.true.) ! -d, --database <path>
        call parser%add('input',     short='i', type=ARG_TYPE_FILE,     required=.true., exist=.true.) ! -i, --input <path>
        call parser%add('quote',     short='q', type=ARG_TYPE_CHAR)    ! -q, --quote <char>
        call parser%add('separator', short='s', type=ARG_TYPE_CHAR)    ! -s, --separator <char>
        call parser%add('dry',       short='D', type=ARG_TYPE_LOGICAL) ! -D, --dry
        call parser%add('verbose',   short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('type',      type_name)
        call parser%get('database',  app%database)
        call parser%get('input',     app%input)
        call parser%get('quote',     app%quote)
        call parser%get('separator', app%separator)
        call parser%get('dry',       app%dry)
        call parser%get('verbose',   app%verbose)

        app%type = dm_type_from_name(type_name)

        ! Validate arguments.
        rc = validate(app)
    end function read_args

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. app%dry) then
            if (.not. dm_string_has(app%database)) then
                call dm_error_out(rc, 'argument --database required')
                return
            end if

            if (.not. dm_file_exists(app%database)) then
                call dm_error_out(rc, 'database ' // trim(app%database) // ' not found')
                return
            end if
        end if

        select case (app%type)
            case (TYPE_NODE, TYPE_SENSOR, TYPE_TARGET, TYPE_OBSERV, TYPE_LOG)
                continue
            case default
                call dm_error_out(rc, 'invalid type')
                return
        end select

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_db_version(.true.)
    end subroutine version_callback
end program dmimport
